### PR TITLE
[Merged by Bors] - feat: some basic quaternion facts

### PR DESCRIPTION
**Pull request by ericrbg**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Quaternion.lean [hunk 1]
```diff
@@ -572,6 +586,18 @@ theorem finrank_eq_four [StrongRankCondition R] : FiniteDimensional.finrank R 
   rw [FiniteDimensional.finrank, rank_eq_four, Cardinal.toNat_ofNat]
 #align quaternion_algebra.finrank_eq_four QuaternionAlgebra.finrank_eq_four
 
+def swapEquiv : ℍ[R,c₁,c₂] ≃ₐ[R] ℍ[R, c₂, c₁] where
+  toFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  invFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  left_inv _ := by simp
+  right_inv _ := by simp
+  map_mul' _ _ := by ext <;>
```
Comment (ericrbg) [inline Mathlib/Algebra/Quaternion.lean:594]: formatting suggestions very welcome

File: Mathlib/Algebra/Quaternion.lean [hunk 2]
```diff
@@ -102,6 +102,10 @@ theorem mk.eta {R : Type*} {c₁ c₂} (a : ℍ[R,c₁,c₂]) : mk a.1 a.2 a.3 a
 
 variable {S T R : Type*} [CommRing R] {c₁ c₂ : R} (r x y z : R) (a b c : ℍ[R,c₁,c₂])
 
+instance : Nonempty ℍ[R, c₁, c₂] := ⟨⟨c₁,c₁,c₁,c₁⟩⟩
```
Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:105]:
```suggestion
instance : Inhabited ℍ[R, c₁, c₂] := ⟨0⟩
```
would be more canonical

File: Mathlib/Algebra/Quaternion.lean [hunk 3]
```diff
@@ -572,6 +585,19 @@ theorem finrank_eq_four [StrongRankCondition R] : FiniteDimensional.finrank R 
...
+/-- There is a natural equivalence when swapping the coefficients of a quaternion algebra. -/
+def swapEquiv : ℍ[R,c₁,c₂] ≃ₐ[R] ℍ[R, c₂, c₁] where
+  toFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  invFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  left_inv _ := by simp
+  right_inv _ := by simp
+  map_mul' _ _ := by ext <;>
+                     simp only [mul_re, mul_imJ, mul_imI, add_left_inj, mul_imK, neg_mul,
+                                neg_add_rev, neg_sub, mk_mul_mk, mul_neg, neg_neg, sub_neg_eq_add]
+                         <;> ring
```
Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:597]: maybe
```suggestion
  map_mul' _ _ := by
    ext
      <;> simp only [mul_re, mul_imJ, mul_imI, add_left_inj, mul_imK, neg_mul, neg_add_rev,
                              neg_sub, mk_mul_mk, mul_neg, neg_neg, sub_neg_eq_add]
      <;> ring
```

File: Mathlib/Algebra/Quaternion.lean [hunk 4]
```diff
@@ -492,6 +496,15 @@ theorem algebraMap_eq (r : R) : algebraMap R ℍ[R,c₁,c₂] r = ⟨r, 0, 0, 0
   rfl
 #align quaternion_algebra.algebra_map_eq QuaternionAlgebra.algebraMap_eq
 
+theorem algebraMap_injective : (algebraMap R ℍ[R,c₁,c₂] : _ → _).Injective :=
+  fun _ _ ↦ by simp [algebraMap_eq]
```
Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:500]: Can you copy this lemma for `Quaternion` too?

File: Mathlib/Algebra/Quaternion.lean [hunk 5]
```diff
@@ -572,6 +585,19 @@ theorem finrank_eq_four [StrongRankCondition R] : FiniteDimensional.finrank R 
   rw [FiniteDimensional.finrank, rank_eq_four, Cardinal.toNat_ofNat]
 #align quaternion_algebra.finrank_eq_four QuaternionAlgebra.finrank_eq_four
 
+/-- There is a natural equivalence when swapping the coefficients of a quaternion algebra. -/
+def swapEquiv : ℍ[R,c₁,c₂] ≃ₐ[R] ℍ[R, c₂, c₁] where
```
Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:589]: In the context of `ℍ`, this doesn't seem natural at all; it treats `i` and `j` specially, without swapping the others!

I suppose this *is* natural for `ℍ[R,c₁,c₂]`, though I believe it is still available as dot notation on `Quaternion`.

Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:589]:
```suggestion
/-- There is a natural equivalence when swapping the coefficients of a quaternion algebra. -/
@[simps]
def swapEquiv : ℍ[R,c₁,c₂] ≃ₐ[R] ℍ[R, c₂, c₁] where
```

Comment (ericrbg) [reply Mathlib/Algebra/Quaternion.lean:589]: I mean, technically it turns `ij` into `ji = -ij` in that special case, I think. Did I mess up the namespacing?

Comment (eric-wieser) [reply Mathlib/Algebra/Quaternion.lean:589]: No, but `Quaternion := QuaternionAlgebra _ _ _` so dot notation naturally falls back on this definition.

Comment (ericrbg) [reply Mathlib/Algebra/Quaternion.lean]: I didn't know dot notation could see through `def`s. Not sure what's best to do in this case. This is obviously not the standard involution...

File: Mathlib/Algebra/Quaternion.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Algebra/Quaternion.lean:105]: I think TC is fast enough when it comes to finding a zero element.

File: Mathlib/Algebra/Quaternion.lean [hunk 6]
```diff
@@ -572,6 +585,21 @@ theorem finrank_eq_four [StrongRankCondition R] : FiniteDimensional.finrank R 
...
+/-- There is a natural equivalence when swapping the coefficients of a quaternion algebra. -/
+@[simps]
+def swapEquiv : ℍ[R,c₁,c₂] ≃ₐ[R] ℍ[R, c₂, c₁] where
+  toFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  invFun t := ⟨t.1, t.3, t.2, -t.4⟩
+  left_inv _ := by simp
+  right_inv _ := by simp
+  map_mul' _ _ := by
+    ext
+      <;> simp only [mul_re, mul_imJ, mul_imI, add_left_inj, mul_imK, neg_mul, neg_add_rev,
```
Comment (jcommelin) [inline Mathlib/Algebra/Quaternion.lean:597]: Question I don't know the answer to: is it conventional to put `<;>` at the beginning of the line? I would have expected it at the end of the previous line.

Comment (ericrbg) [reply Mathlib/Algebra/Quaternion.lean:597]: i've moved it up, but i'm happy for anything as long as it looks remotely sensible

File: Mathlib/Algebra/Quaternion.lean (hunk 2 above)
Comment (ericrbg) [reply Mathlib/Algebra/Quaternion.lean:105]: this Inhabited instance already exists further down the file, L189. Should I just delete this?

Comment (eric-wieser) [reply Mathlib/Algebra/Quaternion.lean:105]: Deleting it sounds good to me.

File: Mathlib/Algebra/Quaternion.lean [hunk 7]
```diff
@@ -102,6 +102,10 @@ theorem mk.eta {R : Type*} {c₁ c₂} (a : ℍ[R,c₁,c₂]) : mk a.1 a.2 a.3 a
 
 variable {S T R : Type*} [CommRing R] {c₁ c₂ : R} (r x y z : R) (a b c : ℍ[R,c₁,c₂])
 
+instance : Nonempty ℍ[R, c₁, c₂] := ⟨⟨c₁,c₁,c₁,c₁⟩⟩
+instance [Subsingleton R] : Subsingleton ℍ[R, c₁, c₂] := (equivTuple c₁ c₂).subsingleton
+instance [Nontrivial R] : Nontrivial ℍ[R, c₁, c₂] := (equivTuple c₁ c₂).surjective.nontrivial
```
Comment (eric-wieser) [inline Mathlib/Algebra/Quaternion.lean:107]: Do we have these instances for `Quaternion`? If not, can you add them?

**Review by eric-wieser — APPROVED** [review]
bors d+

**Pull request closed without merging**